cmake/MPIDynamicHelper.cmake

```diff
@@ -2,19 +2,69 @@
 # Configure support for dynamic MPI initialization
 # =============================================================================
 # ~~~
-# TODO : verify that NRN_ENABLE_MPI_DYNAMIC is valid and determine an include
+# Verify that NRN_ENABLE_MPI_DYNAMIC is valid and determine an include
 # directory for each MPI package building libnrnmpi_<mpipkg>.so.
 # Depending on the MPIs used NRNMPI_INCLUDE_<mpipkg> will be defined.
-# Currently dynamic MPI support is not implemented.
+# A number of other lists needed to construct libnrnmpi_<mpipkg>.so
+# are also constructed. The lists are all parallel in the sense that
+# corresponding elements are related to the same mpi installation.
 # ~~~
 
+set(NRN_MPI_INCLUDE_LIST "" CACHE INTERNAL "" FORCE)
+set(NRN_MPI_LIBNAME_LIST "" CACHE INTERNAL "" FORCE)
+set(NRN_MPI_TYPE_LIST "" CACHE INTERNAL "" FORCE)
+
 if(NRN_ENABLE_MPI)
   if(NRN_ENABLE_MPI_DYNAMIC)
-    if("${NRN_MPI_DYNAMIC}" STREQUAL "")
-      set(NRNMPI_DYNAMICLOAD 1)
-    else()
-      message(FATAL_ERROR "NRN_ENABLE_MPI_DYNAMIC doesn't support multiple MPI yet")
+
+    set(NRNMPI_DYNAMICLOAD 1)
+    # compute the NRN_MPI_INCLUDE_LIST
+    if("${NRN_MPI_DYNAMIC}" STREQUAL "") # use the MPI already found
+      string(REGEX REPLACE "/$" "" foo "${MPI_C_HEADER_DIR}")
+      list(APPEND NRN_MPI_INCLUDE_LIST "${foo}")
+    else() # find the mpi's in the ';' separated list of mpi include folders.
+      foreach(incdir ${NRN_MPI_DYNAMIC})
+        # remove trailing '/' if it has one
+        string(REGEX REPLACE "/$" "" foo "${incdir}")
+        list(APPEND NRN_MPI_INCLUDE_LIST "${foo}")
+      endforeach(incdir)
     endif()
+
+    # compute the NRN_MPI_TYPE_LIST and NRN_MPI_LIBNAME_LIST
+    foreach(idir ${NRN_MPI_INCLUDE_LIST})
+      # ${idir}/mpi.h must exist
+      if(NOT EXISTS "${idir}/mpi.h")
+        message(FATAL_ERROR "${idir}/mpi.h does not exist")
+      endif()
+
+      # ~~~
+      # Only know about openmpi, mpich, and ms-mpi.
+      # ~~~
+      execute_process(COMMAND grep -q "define OPEN_MPI  *1" ${idir}/mpi.h RESULT_VARIABLE result)
+      if (result EQUAL 0)
+        set(type "ompi")
+      else()
+        execute_process(COMMAND grep -q "define MPICH  *1" ${idir}/mpi.h RESULT_VARIABLE result)
+        if (result EQUAL 0)
+          set(type "mpich")
+        else()
+          execute_process(COMMAND grep -q "define MPT_VERSION " ${idir}/mpi.h RESULT_VARIABLE result)
+          if (result EQUAL 0)
+            set(type "mpt")
+          else()
+            execute_process(COMMAND grep -q "define MSMPI_VER " ${idir}/mpi.h RESULT_VARIABLE result)
+            if (result EQUAL 0)
+              set(type "msmpi")
+            else()
+              # Perhaps rather set the type to "unknown" and check for no duplicates?
+              message(FATAL_ERROR "${idir}/mpi.h does not seem to be either openmpi, mpich, mpt or msmpi")
+            endif()
```

Comment:
`nitpick` : is it possible to avoid `multi-level nested if-else-if` blocks?  for example:

```
set(MPI_FOUND FALSE)
execute_process(COMMAND grep -q "define ...")
if (result EQUAL 0 AND NOT MPI_FOUND)
  set(MPI_FOUND TRUE)
endif
execute_process(COMMAND grep -q "define ...")
if (result EQUAL 0 AND NOT MPI_FOUND)
endif
```

Reply:
That would do the trick (but don't forget to set the type)  I had trouble imagining both doing the execute_process and checking the result within a sequence of  if...else if. Now that there are more than three
cases, it is a shame to nest even further.  And it slightly offends me to do all execute process when the first one is almost always successful (openmpi). How about using the concept of function or macro to do something like
```
if ( mpitype("define OPEN_MPI  *1")
  set(type, "ompi")
elseif (mpitype("define MPICH  *1", type)
  set(type, "mpich")
elseif (mpitype("define MSMPI_VER)
  set(type, "msmpi")
else()
  #fatal error
endif()
```
then it becomes clear how to add new types.

Reply:
this seems much cleaner if this can be done!

Reply:
It appears there is no such thing in cmake as a function return value. So we are stuck with
a nested if or execute_process for every mpi implementation. At least the pattern for nesting is fairly clear in MPIDynamicHelper.cmake .  What do you think about merging as is @pramodk

Reply:
I am fine with this. We will see  if there is an alternative later.